types/basetypes/dynamic_type.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// DynamicTypable extends attr.Type for dynamic types. Implement this interface to create a custom DynamicType type.
+type DynamicTypable interface {
+	attr.TypeWithDynamicValue
+
+	// ValueFromDynamic should convert the DynamicValue to a DynamicValuable type.
+	ValueFromDynamic(context.Context, DynamicValue) (DynamicValuable, diag.Diagnostics)
+}
+
+var _ DynamicTypable = DynamicType{}
+
+// DynamicType is the base framework type for a dynamic. Static types are always
+// preferable over dynamic types in Terraform as practitioners will receive less
+// helpful configuration assistance from validation error diagnostics and editor
+// integrations.
+//
+// DynamicValue is the associated value type and, when known, contains a concrete
+// value of another framework type. (StringValue, ListValue, ObjectValue, MapValue, etc.)
+type DynamicType struct{}
+
+// ApplyTerraform5AttributePathStep applies the given AttributePathStep to the type.
+func (t DynamicType) ApplyTerraform5AttributePathStep(step tftypes.AttributePathStep) (interface{}, error) {
+	// MAINTAINER NOTE: Based on dynamic type alone, there is no alternative type information to return related to a path step.
+	// However, it is possible for a dynamic type to have an underlying value of a list, set, map, object, or tuple
+	// that will have corresponding path steps for an element type.
+	//
+	// Since the dynamic type has no reference to the underlying value, we just return the dynamic type which can be used
+	// to determine the correct attr.Value from `(DynamicType).ValueFromTerraform`.
+	return t, nil
+}
+
+// Equal returns true if the given type is equivalent.
+//
+// Dynamic types do not contain a reference to the underlying `attr.Value` type, so this equality check
+// only asserts that both types are DynamicType.
+func (t DynamicType) Equal(o attr.Type) bool {
+	_, ok := o.(DynamicType)
+
+	return ok
+}
+
+// String returns a human-friendly description of the DynamicType.
+func (t DynamicType) String() string {
+	return "basetypes.DynamicType"
+}
+
+// TerraformType returns the tftypes.Type that should be used to represent this type.
+func (t DynamicType) TerraformType(ctx context.Context) tftypes.Type {
+	return tftypes.DynamicPseudoType
+}
+
+// ValueFromDynamic returns a DynamicValuable type given a DynamicValue.
+func (t DynamicType) ValueFromDynamic(ctx context.Context, v DynamicValue) (DynamicValuable, diag.Diagnostics) {
+	return v, nil
+}
+
+// ValueFromTerraform returns an attr.Value given a tftypes.Value. This is meant to convert
+// the tftypes.Value into a more convenient Go type for the provider to consume the data with.
+func (t DynamicType) ValueFromTerraform(ctx context.Context, in tftypes.Value) (attr.Value, error) {
+	if !in.IsKnown() {
+		return NewDynamicUnknown(), nil
+	}
+
+	if in.IsNull() {
+		return NewDynamicNull(), nil
+	}
```

Comment:
Reminder note about a minor bug you discovered and we chatted about: Terraform configurations can cause a situation where a dynamic type constraint is actually "known" in the sense that Terraform's type system has refined the value into a concrete type unknown/null. You had found an excellent situation with taking a configuration that returned an unknown value, but of a concrete type, and passing that into a dynamic attribute. e.g.

```terraform
resource "random_shuffle" "test" {
  input = ["one", "two"]
  result_count = 2
}

resource "examplecloud_thing" "test" {
  # on plan, this will come in as an unknown list of strings
  # (since the random provider cannot yet do its calculations during plan)
  dynamic = random_shuffle.test.result
}
```

Which coerced Terraform into reporting a roundtrip data consistency issue during plan:

```
│ Error: Provider produced invalid plan
│ 
│ Provider "registry.terraform.io/austinvalle/sandbox" planned an invalid value for examplecloud_thing.this.dynamic: planned value cty.DynamicVal does not match config value cty.UnknownVal(cty.List(cty.String)).
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

Since with this current logic, it would cause the already "known-ness" to be lost which Terraform requires you to roundtrip back. Your proposed fix here of checking `in.Type().Is(tftypes.DynamicPseudoType)` will enable the type system to preserve the exact situation better, although it may feel a little semantically awkward of first. Dynamic types are more about the type constraint itself than the values themselves in this case, so it is valid to say the dynamic value is "known" even though its underlying value might be unknown or null. There are cases where the dynamic value coming in will have no additional type information, so that's when the dynamic type itself is null or unknown.

I hope we can write some decent words about this in the documentation because it may be confusing at first, but the type system correctness here (and necessity to prevent Terraform errors) means we need to implement those sorts of semantics and expose them to developers trying to implement this advanced use case type. 😄

Reply:
Check out 90c4253, added the fix along with some tests and an attempt at documentation 😆

Reply:
Fun fact, I also need to update the planmodifiers like `UseStateForUnknown` now 😆

I'm sure there will be other places that need to be updated

Reply:
A simple way to test the null variant of this:

```
resource "examplecloud_thing" "this" {
  # on plan, this will come in as a null of type string
  dynamic = var.null_typed
}

variable "null_typed" {
  type    = string
  default = null
}

```

Reply:
From out-of-band conversation, we had discussed potentially introducing additional method(s) for dynamic values that could return a boolean for checking the odd case of nil value and null underlying value state together. `IsNull()` isn't right because folks should be able to distinguish between the dynamic "container" value state and the underlying value state separately, so maybe something (potentially better named than my suggestion here) like `HasUnderlyingNull()`. It can be documented as a helper to prevent potential panics or something. 👍

Reply:
Thanks for the summary @bflad! I just added these methods, any comments on potential naming is appreciated 😆 

9c25457 ⚡